### PR TITLE
156 event handlers support arbitrary format

### LINKED ISSUE #156: Event-handlers support arbitrary format
**Is your feature request related to a problem? Please describe.**
All our existing ignite Event-Handlers only support fixed array data. Need to support arbitrary format model input and output with **batch_transform** and **output_transform**, especially for dict format.

**Describe the solution you'd like**
Similar to ignite trainer or evaluator class.

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #156 .

### Description
This PR updated all the event handlers' API and corresponding examples for arbitrary input format.

### Status
**Ready**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### CODE REVIEW

File: examples/unet_inference_3d_array.py [hunk 1]
```diff
@@ -75,11 +77,10 @@ def _sliding_window_processor(_engine, batch):
...
 
-# checkpoint_handler = ModelCheckpoint('./', 'net', n_saved=10, save_interval=3, require_empty=False)
-# infer_engine.add_event_handler(event_name=Events.EPOCH_COMPLETED, handler=checkpoint_handler, to_save={'net': net})
-
-SegmentationSaver(output_path='tempdir', output_ext='.nii.gz', output_postfix='seg').attach(infer_engine)
-CheckpointLoader(load_path='./net_checkpoint_9.pth', load_dict={'net': net}).attach(infer_engine)
+# for the arrary data format, assume the 3rd item of batch data is the meta_data
+SegmentationSaver(output_path='tempdir', output_ext='.nii.gz', output_postfix='seg',
+                  batch_transform=lambda x: x[2]).attach(infer_engine)
+CheckpointLoader(load_path='./runs/net_checkpoint_120.pth', load_dict={'net': net}).attach(infer_engine)
```
Comment (wyli) [inline examples/unet_inference_3d_array.py]: perhaps add a comment here about how to get `./runs/net_checkpoint_120.pth`?

File: monai/utils/sliding_window_inference.py [hunk 2]
```diff
@@ -49,7 +49,7 @@ def sliding_window_inference(inputs, roi_size, sw_batch_size, predictor, device)
     # in case that image size is smaller than roi size
     image_size = tuple(max(image_size[i], roi_size[i]) for i in range(num_spatial_dims))
     inputs = ImageEndPadder(roi_size, 'constant')(inputs)  # in np array
-    inputs = ToTensor()(inputs)
+    inputs = convert_tensor(ToTensor()(inputs), device, False)
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/utils/sliding_window_inference.py]: Fixed in the latest patch.
Thanks.

File: examples/unet_inference_3d_array.py (hunk 1 above)
Comment (Nic-Ma) [reply examples/unet_inference_3d_array.py]: Fixed in the latest patch.
Thanks.

File: examples/unet_inference_3d_array.py [hunk 3]
```diff
@@ -21,6 +21,9 @@
 from ignite.engine import Engine
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply examples/unet_inference_3d_array.py]: Updated in the latest patch.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Thanks for your review!
I updated code according to the comments, could you please help review it again?
Thanks.

File: monai/handlers/segmentation_saver.py [hunk 4]
```diff
@@ -88,24 +91,24 @@ def _create_file_basename(postfix, input_file_name, folder_path, data_root_dir="
...
         """
-        This method assumes:
-            - 3rd output of engine.state.batch is a meta data dict, and have the keys:
+        This method assumes to have the keys:
             'filename_or_obj' -- for output file name creation
             and optionally 'original_affine', 'affine' for data orientation handling.
             - output file datatype from `engine.state.output.dtype`.
         """
+        meta_data = self.batch_transform(engine.state.batch)
         meta_data = engine.state.batch[2]  # assuming 3rd output of input dataset is a meta data dict
```
Comment (wyli) [inline monai/handlers/segmentation_saver.py]: need to remove line 100

File: examples/unet_segmentation_3d_array.py [hunk 5]
```diff
@@ -148,7 +148,7 @@ def log_training_loss(engine):
 
 # Add stats event handler to print validation stats via evaluator
 logging.basicConfig(stream=sys.stdout, level=logging.INFO)
-val_stats_handler = StatsHandler()
+val_stats_handler = StatsHandler(output_transform=lambda output: (None, None))
```
Comment (wyli) [inline examples/unet_segmentation_3d_array.py]: this is not very intuitive, is the intention blocking iteration default output here?

File: monai/handlers/segmentation_saver.py (hunk 4 above)
Comment (wyli) [reply monai/handlers/segmentation_saver.py]: could have this docstring: this method assumes `self.batch_transform` will extract meta data from the input batch, that is, `meta_data = self.batch_transform(engine.state.batch)`

File: monai/handlers/stats_handler.py [hunk 6]
```diff
@@ -125,8 +134,8 @@ def _default_iteration_print(self, engine: Engine):
             engine (ignite.engine): Ignite Engine, it can be a trainer, validator or evaluator.
 
         """
-        prints_dict = engine.state.metrics
-        if not prints_dict:
+        loss = self.output_transform(engine.state.output)[1]
+        if loss is None:
```
Comment (wyli) [inline monai/handlers/stats_handler.py]: perhaps don't print it if state.output is high dimensional data?

File: monai/handlers/segmentation_saver.py (hunk 4 above)
Comment (Nic-Ma) [reply monai/handlers/segmentation_saver.py]: fixed in the latest patch.
Thanks.

File: examples/unet_segmentation_3d_array.py (hunk 5 above)
Comment (Nic-Ma) [reply examples/unet_segmentation_3d_array.py]: fixed in the latest patch.
Thanks.

File: monai/handlers/stats_handler.py (hunk 6 above)
Comment (Nic-Ma) [reply monai/handlers/stats_handler.py]: fixed in the latest patch.
Thanks.

**Pull request merged**